File: PixiJS.Engine/TankGame/src/KFramework.PixiJS/Tool/RandomTool.ts
/**
 * 随机数工具（移植自 MonoGame 版 RandomTool）。
 *
 * 与 C# 版的差异：
 * - C# 用 `[ThreadStatic]` 的 Random 实例，JS 是单线程的，直接用 `Math.random()`
 * - C# 的 `RandomInt64` / `RandomUInt64` 没有移植：JS 的 number 是双精度浮点，
 *   安全整数范围只有 ±2^53，装不下 64 位随机整数，真需要请用 BigInt 自行扩展
 * - C# 的 `RandomArrayIndex` / `RandomInt32` / `RandomUInt32` 都归并到 `randomInt`
 * - 取值区间与 C# 的 `Random.Next(x, y)` 一致，都是 `[min, max)`
 * - `getIndexByRate` 修掉了 C# 版的边界问题（原版取不到最后一个区间的尾部权重）
 */
export class RandomTool {
  /** `[min, max)` 的随机整数 */
  public static randomInt(min: number, max: number): number {
    console.assert(
      min <= max,
      `RandomTool.randomInt: min(${min}) > max(${max})`,
    );
    return min + Math.floor(Math.random() * (max - min));
  }

  /** `[min, max)` 的随机浮点数 */
  public static randomFloat(min: number, max: number): number {
    console.assert(
      min <= max,
      `RandomTool.randomFloat: min(${min}) > max(${max})`,
    );
    return min + Math.random() * (max - min);
  }

  /**
   * 按权重随机取下标。
   * @param rates 每项权重，非正数按 0 处理；总权重 <= 0 时返回 -1
   */
  public static getIndexByRate(rates: readonly number[]): number {
    let total = 0;
    for (const rate of rates) {
      total += rate;
    }
    if (total <= 0) return -1;

    let target = RandomTool.randomInt(0, total);
    for (let i = 0; i < rates.length; i++) {
      target -= rates[i];
      if (target < 0) return i;
    }
    return rates.length - 1;
  }
}

File: KFramework.WebGame/KFramework.MonoGameEx/KFramework.MonoGame/Tools/RandomTool.cs
using System;
using System.Runtime.CompilerServices;
using System.Security.Cryptography;

namespace KFramework.MonoGame
{
    public static class RandomTool
    {
        [ThreadStatic]
        private static Random m_Random;

        private static Random RandomGenerator
        {
            get
            {
                if (m_Random == null)
                {
                    m_Random = new Random(RandomNumberGenerator.GetInt32(int.MaxValue));
                }
                return m_Random;
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int RandomArrayIndex(int x, int y)
        {
            return RandomGenerator.Next(x, y);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static long InnerRandomInt64(long x, long y)
        {
            return (long)(x + RandomGenerator.NextDouble() * (y - x));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static ulong InnerRandomUInt64(ulong x, ulong y)
        {
            return (ulong)(x + RandomGenerator.NextDouble() * (y - x));
        }

        public static int RandomInt32(int x, int y)
        {
            PrintTool.Assert(x >= int.MinValue);
            PrintTool.Assert(y <= int.MaxValue);
            PrintTool.Assert(x <= y);
            return (int)InnerRandomInt64(x, y);
        }

        public static uint RandomUInt32(uint x, uint y)
        {
            PrintTool.Assert(x >= uint.MinValue);
            PrintTool.Assert(y <= uint.MaxValue);
            PrintTool.Assert(x <= y);
            return (uint)InnerRandomInt64(x, y);
        }

        public static ulong RandomUInt64(ulong x, ulong y)
        {
            PrintTool.Assert(x >= ulong.MinValue);
            PrintTool.Assert(y <= ulong.MaxValue);
            PrintTool.Assert(x <= y);
            return InnerRandomUInt64(x, y);
        }

        public static long RandomInt64(long x, long y)
        {
            PrintTool.Assert(x >= long.MinValue);
            PrintTool.Assert(y <= long.MaxValue);
            PrintTool.Assert(x <= y);
            return InnerRandomInt64(x, y);
        }

        public static int GetIndexByRate(int[] mRateList)
        {
            int nSumRate = 0;
            foreach (var nRate in mRateList)
            {
                nSumRate = nSumRate + nRate;
            }

            int nTempTargetRate = nSumRate + 1;
            if (nSumRate >= 1)
            {
                nTempTargetRate = RandomInt32(1, nSumRate);
            }

            int nTempRate = 0;
            int nTargetIndex = -1;
            for (int i = 0; i < mRateList.Length; i++)
            {
                nTempRate = nTempRate + mRateList[i];
                if (nTempRate >= nTempTargetRate)
                {
                    nTargetIndex = i;
                    break;
                }
            }

            return nTargetIndex;
        }
    }
}


